File: Java/src/MinimumStack/MinStack.java
package MinimumStack;

import java.util.*;

class MinStack {
    // Linked list is slower than ArrayDeque
    Deque<Integer> mainStack = new ArrayDeque<Integer>();
    Deque<Integer> minStack = new ArrayDeque<Integer>();

    public void push(int x) {
        mainStack.push(x);
        if ((!minStack.isEmpty() && minStack.peek() >= x) || minStack.isEmpty()) {
            minStack.push(x);
        }
    }

    public void pop() {
        assert(!mainStack.isEmpty());
        int peek = mainStack.pop();
        if (!minStack.isEmpty() && peek == minStack.peek()) {
            minStack.pop();
        }
    }

    public int top() {
        assert(!mainStack.isEmpty());
        return mainStack.peek();
    }

    public int getMin() {
        assert(!minStack.isEmpty());
        return minStack.peek();
    }
}

File: Python/MinStack/MinStack.py
class MinStack:
    def __init__(self):
        self.mainStack = []
        self.minStack = []

    # @param x, an integer
    # @return an integer
    def push(self, x):
        self.mainStack.append(x)
        if (self.minStack and self.minStack[-1] >= x) or not self.minStack:
            self.minStack.append(x)

    # @return nothing
    def pop(self):
        peek = self.mainStack.pop()
        if self.minStack and self.minStack[-1] == peek:
            self.minStack.pop()

    # @return integer
    def top(self):
        assert self.mainStack
        return self.mainStack[-1]

    # @return integer
    def getMin(self):
        assert self.minStack
        return self.minStack[-1]

